doc/config_reload/config_reload_enhancement.md

```diff
@@ -0,0 +1,122 @@
+```
+Feb 8 21:13:46.695474 sonic ERR syncd#SDK: :- threadFunction: time span WD exceeded 30955 ms for create:SAI_OBJECT_TYPE_SWITCH:oid:0x21000000000000
+Feb 8 21:13:46.695474 sonic ERR syncd#SDK: :- logEventData: op: create, key: SAI_OBJECT_TYPE_SWITCH:oid:0x21000000000000
+Feb 8 21:13:46.695474 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_INIT_SWITCH: true
+Feb 8 21:13:46.695474 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_FDB_EVENT_NOTIFY: 0x55dabccda690
+Feb 8 21:13:46.695474 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_PORT_STATE_CHANGE_NOTIFY: 0x55dabccda6a0
+Feb 8 21:13:46.695474 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_BFD_SESSION_STATE_CHANGE_NOTIFY: 0x55dabccda6b0
+Feb 8 21:13:46.695525 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_SWITCH_SHUTDOWN_REQUEST_NOTIFY: 0x55dabccda6c0
+Feb 8 21:13:48.653822 sonic ERR syncd#SDK: :- setEndTime: event 'create:SAI_OBJECT_TYPE_SWITCH:oid:0x21000000000000' took 32913 ms to execute
+Feb 8 21:13:48.654048 sonic ERR syncd#SDK: :- logEventData: op: create, key: SAI_OBJECT_TYPE_SWITCH:oid:0x21000000000000
+Feb 8 21:13:48.654230 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_INIT_SWITCH: true
+Feb 8 21:13:48.654405 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_FDB_EVENT_NOTIFY: 0x55dabccda690
+Feb 8 21:13:48.654971 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_PORT_STATE_CHANGE_NOTIFY: 0x55dabccda6a0
+Feb 8 21:13:48.655165 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_BFD_SESSION_STATE_CHANGE_NOTIFY: 0x55dabccda6b0
+Feb 8 21:13:48.655338 sonic ERR syncd#SDK: :- logEventData: fv: SAI_SWITCH_ATTR_SWITCH_SHUTDOWN_REQUEST_NOTIFY: 0x55dabccda6c0
+```
+This problem is solved during the reboot by selectively delaying non critical services like snmp, telemetry, etc,. using systemd timers with fixed delay. This approach is not event driven and may not be optimal during config reload.
+In this new design, the non critical services will be started only after the ports are initialized using an event driven approach.
+
+### Requirements
+
+Primary requirements for sequencing the config reload are
+- Immediately restart the critical services during config reload.
+- The non critical should be started only after all the ports are initialized.
+- Services can be configured to be started immediately or delayed. This can be using a field in FEATURE table.
+- The existing timers should be removed by this event driven approach.
+- This flow is applicable in case of all reboots (warm/fast/cold) as well as config reload.
+
+
+### High-Level Design
+Currently hostcfgd controls the services based on the feature table. The feature table has a specific field 'has_timer' for the non essential services which needs to be delayed during the reboot flow. This field will be now replaced by new field called "delayed". These services will controlled by hostcfgd.

```

Comment:
what if just hostcfgd is restarted, since it's now subscribing for portinitdone?

Reply:
Hostcfgd will restart the delayed services on seeing the portinitdone. If the services are already started this will be a no op. Today hostcfgd does the same as we already start many services before hostcfgd starts and we encounter service twice. The 2nd time through hostcfgd is a no-op.